api/dataprocessor.go

```diff
@@ -226,52 +228,52 @@ func (s *Server) getTargets(ctx context.Context, ss *models.StorageStats, reqs [
 // getTargetsRemote issues the requests on other nodes
 // it's nothing more than a thin network wrapper around getTargetsLocal of a peer.
 func (s *Server) getTargetsRemote(ctx context.Context, ss *models.StorageStats, remoteReqs map[string][]models.Req) ([]models.Series, error) {
-	responses := make(chan getTargetsResp, len(remoteReqs))
-	rCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	wg := sync.WaitGroup{}
-	wg.Add(len(remoteReqs))
+
+	allPeers, err := cluster.MembersForSpeculativeQuery()
+	if err != nil {
+		return nil, err
+	}
+
+	requiredPeers := make(map[int32][]cluster.Node)
+	shardReqs := make(map[int32][]models.Req)
 	for _, nodeReqs := range remoteReqs {
-		log.Debugf("DP getTargetsRemote: handling %d reqs from %s", len(nodeReqs), nodeReqs[0].Node.GetName())
-		go func(reqs []models.Req) {
-			defer wg.Done()
-			node := reqs[0].Node
-			buf, err := node.Post(rCtx, "getTargetsRemote", "/getdata", models.GetData{Requests: reqs})
-			if err != nil {
-				cancel()
-				responses <- getTargetsResp{nil, err}
-				return
-			}
-			var resp models.GetDataRespV1
-			_, err = resp.UnmarshalMsg(buf)
-			if err != nil {
-				cancel()
-				log.Errorf("DP getTargetsRemote: error unmarshaling body from %s/getdata: %q", node.GetName(), err.Error())
-				responses <- getTargetsResp{nil, err}
-				return
-			}
-			log.Debugf("DP getTargetsRemote: %s returned %d series", node.GetName(), len(resp.Series))
-			ss.Add(&resp.Stats)
-			responses <- getTargetsResp{resp.Series, nil}
-		}(nodeReqs)
+		shardID := nodeReqs[0].Node.GetPartitions()[0]
+		requiredPeers[shardID] = allPeers[shardID]
+		// TODO - Sometimes two peers from the same shard have different responses here...very strange
```

Comment:
what kind of different responses are you getting exactly? different number or set of series? different values or timestamps?
have you noticed a way to reproduce?

should this code be considered alpha? are you running this in production?

Reply:
in your other comment you wrote:

>  2 peers of the same shard group are marked as having some of the data.

assuming that's the issue here, can you elaborate? both have only partial data? "marked" how/where?

Reply:
Ah, yeah that wasn't clear. The initial code I wrote assumed that the `remoteReqs` map would have one peer from each shard had data. But what I actually found is that there would sometimes be multiple peers from the same shard each in `remoteReqs` and some reqs for each of them. I expect this is due to multiple rounds of `find_by_tag` requests picking different peers.

Reply:
looking into this code a bit more, trying to understand the issue.
1) `Server.executePlan` calls s.clusterFindByTag / s.findSeries. both rely on `queryAllShardsGeneric` which uses `cluster.MembersForSpeculativeQuery()`. Which means for the series coming back from the find calls, we assign a peer node to them that could be either of the replicas for each shard, though all series would have the same one, because when we query speculatively, we either take the response from one peer *or* its replica, not a mixture of both [*]

2) `Server.executePlan` transforms the series from the finds into models.Req, keeping the cluster node assocations
 get turned into models.Req-peer assocations
3) still in `Server.executePlan`, we call `s.getTargets()` which calls `s.getTargetsRemote`

If this commit really fixes a bug, it means we have different requests mapped to different peer names that host the same shards, e.g. different replicas. Perhaps the assumption marked `[*]` above is wrong, though looking at the code, it seems to check out.

I'm happy to help with digging in here.

Reply:
Seems `queryPeers` seems solid wrt only returing one replica per shard.
Do you only see this when you using multiple queries/patterns/SeriesByTag statements?
If yes: the most obvious explanation is what you allude to: `executePlan` runs the find functions repeatedly and each time it gets a different set of replicas from each shard (by design). In this case, I would consider this case closed and let's remove the TODO.

if no, we need to keep digging.

Reply:
Hmm, I don't recall anymore. I do know that this code handled both cases and in the cases I saw the reqs were non-overlapping (which would explain why there was no issues)